Integrate `ukplat_memregion` into `Xen` `x86`, `AArch64` `KVM` and `linuxu`

mogasergiu opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): `x86_64`, `arm64`
 - Platform(s): `kvm`, `xen`, `linuxu`
 - Application(s):


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Integrate `ukplat_memregion_*` into `KVM` `ARM64` and `Xen` `x86` and `linuxu` to ease memory region management. Furthermore, add an `ARM64` specific intermediary page-fault handler to `ukvmem` to enable `ARM64` to have access to this library and better virtual memory management.

Note: This PR depends on #772 and since it has not been merged it, it contains its commits.

### NOTE
This is part of a larger set of Pull Requests. To make things easier and to ensure that things build as expected, it is recommended that testing is done based on #912, which includes everything. The splitting has been done to ease the review process.

mogasergiu committed:
plat/common: Make the amount of Memory Region Descriptors configurable

Sice `bootinfo` statically allocates space in the final image to store
Memory Region Descriptors at runtime and their actual amount is highly
variable depending on the platform, allow the user to configure a
precise maximum amount of such descriptors.

This allows flexibility in the size of the final image. One can check
at runtime whether `-ENOMEM` is returned by the memory region insertion
function or not and adjust `CONFIG_UKPLAT_MEMREGION_MAX_COUNT` accordingly.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm/x86: Fix `bootmemory_palloc` keeping empty memory region

Whenever a memory region allocation request done through
`bootmemory_palloc` would require a length equal to one of the
available memory regions, a new, equivalent, memory region would
be created and the original one would be left empty, thus breaking
the memory allocator.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/common: Add `dtb` field `bootinfo` structure

Add a pointer to the devicetree blob's address. Since it is a
frequently accessed resource, it comes in handy to be able to
quickly have access to it.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/common: Generalize `bootmemory_palloc`

Since `bootmemory_palloc` can be used by any platform and
architecture and can come in handy when wanting to allocate a memory
region descriptor on the spot, that is not part of the original
memory map, move it to a more general location.

Furthermore, rename the function to `ukplat_memregion_alloc` to make
its usage-case more intuitive.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/common: Add a memory region descriptor coalescing method

Implement a function that, given a list of memory region descriptors,
coalesces them based on a priority. The lowest priority is that of
the free memory regions, while the highest is that of the reserved
memory regions. This priority is required whenever it comes to
splitting memory regions that overlap. Thus, a memory region whose
priority is higher, gets to keep its overlapping fragment, while the
other is either split or reduced.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm/x86: Coalesce the memory region descriptor list

Ensure that there are no overlapping memory regions in the memory
region descriptor list by coalescing them after finishing the
insertion phase of the memory regions reported through the multiboot
protocol.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/linuxu: Integrate `bootinfo` into `linuxu` builds

Add into the build process the sources required for integrating
the `struct ukplat_bootinfo` structure into `linuxu` images, as well
as the call to `build_bootinfo` method.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/xen: Integrate `bootinfo` into `xen` builds

Add into the build process the sources required for integrating
the `struct ukplat_bootinfo` structure into `xen` images, as well
as the call to `build_bootinfo` method.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat: Move `build_bootinfo` to a platform common place

Since `build_bootinfo` can be used for other platforms as well, move
it to a more platform agnostic location.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/linuxu: Move `initrd` and `heap` initialization to `setup.c`

Initialize the `heap` and `initrd` as soon as possible and move the
respective functions to `setup.c`, as that is the only place where
they are to be used.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/linuxu: Manage memory regions through `ukplat_memregion` API's

Switch to using the generic `ukplat_memregion` API's since they are
more dynamic and flexible.

Thus, remove conflicting definitions of `ukplat_memregion` from
`memory.c`, in favor of those from `platform/common`.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

s into linuxu use memregion

mogasergiu committed:
plat/xen: Manage memory regions through `ukplat_memregion` API's

Switch to using the generic `ukplat_memregion` API's since they are
more dynamic and flexible.

Thus, remove conflicting definitions of `ukplat_memregion` from
`memory.c`, in favor of those from `platform/common`.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm: Move `ukplat_memregion` API's to platform common location

Since these API's can be used on any platform and any architecture
subsystem that integrates `struct ukplat_bootinfo` structure, move
their definitions to a platform common place in the tree.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/common/arm: Make `lcpu_arch_jump_to` not SMP dependent

Since `lcpu_arch_jump_to` can be used with and without
`CONFIG_HAVE_SMP`, move it to a place where the preprocessing
phase does not remove the function from the final image.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm/arm: Make the earliest bootstack be part of the image

In order to make it easier for the paging API to not invalidate the
stack, make it part of the image (.bss), so that it is automatically
ignored.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/common: Add `flags` field to `ukplat_memregion_alloc`

Increase `ukplat_memregion_alloc`'s memory region allocation
flexibility by giving the caller the ability to customize
the memory region's flags.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/common: Make `ukplat_memregion_alloc` aware of memory holes

In case we are somehow able to allocate memory from an in-image
memory hole, make sure we do not give it the `UKPLAT_MEMRF_MAP`
flag, since it is already mapped, which causes the paging
initialization phase to exit with an `-EEXIST` error code.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat: Integrate `ukplat_memregion` API's into `ARM64` builds

Rely on the generic `ukplat_memregion` API's for memory region
management and drop `_libkvmplat_cfg` usage in favor of the
`bootinfo` getters.

Now that `ukplat_memregion`'s are usable on `ARM64`, proper
paging support can be used instead of the overly-complicated
memory re-adjustment that was previously done in `_init_dtb_mem`.
Instead, we rely on `ukplat_memregion_alloc` for proper memory
region management.

Another important, notable change, is that of changing
`arm64_bpt_l0_pt0` to contain only one, unified, entry of `pte_fill`.
This reduced page permission granularity allows us further flexibility
in where the Kernel is placed so that we are truly position independent.
One may be right in saying that this looks not so secure, having all that
memory mapped with all permissions, however this is not as important when
the two existing possible use-cases are taken into consideration:
- Paging API is disabled and we will either way have wrong permissions
everywhere anyway
- Paging API is enabled and this area will be unmapped anyway and new
proper permissions will be used instead
Therefore, until the time comes when we manage to dynamically get rid
of the static boot page tables, we will align ARM with x86 and have the
static boot page tables all mapped with all permissions w.r.t. to
Kernel placement.

Furthermore, now that ARM64 can also use the heap base
`CONFIG_LIBUKBOOT_HEAP_BASE` configuration option, drop the x86
dependency.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm: Drop no longer used `libkvmplat_cfg` structure

Now that we rely on `ukplat_memregion` generic API's, we have no
more need for KVM configuration specific structures when it comes
to memory regions.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/linuxu: Drop no longer used `liblinuxuplat_opts` structure

Now that we rely on `ukplat_memregion` generic API's, we have no
more need for linuxu configuration specific structures when it comes
to memory regions.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm/x86: Mark the thrashed command-line as a kernel resource

Since this buffer is only meant for `ukplat_entry_argp`, which will
thrash it as a result of parsing it to obtain `argc`/`argv`, mark it
as a kernel related memory region instead (`UKPLAT_MEMRT_KERNEL`).

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat: Move paging initialization to a platform common location

Since the `paging_init` function of the x86 KVM subsystem is written
in a generic manner, it can be reused by other architectures or
platforms. Thus, move it to a platform generic location and rename
it with a more intuitive name to align it with the other function
declarations of `paging.c`.

Furthermore, only call `ukplat_pt_set_active` at the very end, to avoid
unnecessary `TLB` flushes.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/common/paging.c: Flush TLB only if we are modifying current PT

Flushing the TLB introduces an unnecessary performance penalty if we
are modifying a page table that is different from our current one.
Therefore, ensure that the TLB is flushed only if the page table we
are currently modifying is the same as the one we currently have
active.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

jschlumpp commented on support/scripts/elf64_to_32.py:
Out of curiosity: Is there a reason to use this over `objcopy --output-target elf32-i386`? 
But I have maybe an even better idea: maybe the script could *prepend* the 32-bit ELF header + multiboot header to the image. Then we would have the proper 64-bit ELF header in the address space and we could also load the symbol table in the future for symbolization of crash back traces.

mogasergiu replied:
This is related to the PIE PR linked in the description. I thought that a custom script offered more control, as objcopy does some things differently, for example it re-adjusts the offset to the Program Headers, while the script leaves everything as it is. Besides, unrelated to ELF32, but I had some issues whenever I would use it with `--target efi-app-x86_64 --subsystem=10` on the 64-bit PIE and I thought that having something I control would be better in general, as I found `objcopy` to sometimes prove unreliable when converting formats.

I agree, that is a great idea, I will change this so it prepends it instead :) good point.

mogasergiu replied:
One problem that I have just thought of though: this implies building the code as if it  were aware of the elf64 header. So I guess this means that we should include it into a PT_LOAD program header?

mogasergiu replied:
Or better yet, perhaps make the script change the in-file offset/size of the first PT-LOAD to also include the elf64 header.

jschlumpp replied:
Yes, basically the kernel would end up in configuration as if the loader would have loaded the original ELF64.
Actually using ELF64 for a x86_64 binary would be the more sensible behavior (Currently it doesn't matter because there is no code in staging that reads the ELF header). Using ELF32 was more of a hack when I removed the multiboot `AOUT_KLUDGE` and `--omagic` flag.
It's not necessary to use the script to insert a new segment for the ELF header (including program headers). Adjusting the linker script to include them in the first PT_LOAD segment is enough.

mogasergiu commented:
I will bring some more changes to this PR in the following 2 days (it's still a draft). Afterwards, hopefully, I will mark it as ready for review, together with the linked PIE PR.

mogasergiu committed:
plat/common: Implement `ukplat_bootinfo_fdt_setup`

Add a method that updates the `ukplat_bootinfo` structure with
relevant information extracted from a `Devicetree` passed as
argument.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/common: Implement insert method for the legacy x86 high memory

Write a short inline function to insert the x86 legacy high memory
area right below the first 1MiB, in the `0xA0000` to `0x100000` range.

Preferably, since this area is very low in memory, it should be added
before we start building the memory region list to save some iterations.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/common: Implement method to dynamically assign `SIPI Vector`

Add an inline function that simply allocates a reserved memory region
meant for the `SIPI Vector`. If this function fails we wouldn"t have
been able to do this properly with the old way anyway (hardcoded
`SIPI Vector` at `0x8000`).

This should only be called after the memory region list has been
fully built and coalesced.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm/x86: Hardcode a memory region descriptor for unmappings

Add a memory region descriptor to indicate the range we are meant to
unmap when paging is initialized and we are getting rid of the
initial static boot page tables.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat: Generalize memory initialization

Re-define `mem_init` into `ukplat_mem_init` that does the same thing,
but does not assume what regions to keep or unmap. Instead, it relies
on the platform/architecture defined unmap memory region to do this.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm/arm: Hardcode a memory region descriptor for unmappings

Add a memory region descriptor to indicate the range we are meant to
unmap when paging is initialized and we are getting rid of the
initial static boot page tables.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm/x86: Keep `Multiboot1` memory regions below `1 MiB`

Keep the `e820` enries below `1 MiB` that we were passed through
`Multiboot1` and rely on the unmappings memory region descriptor to
tell us later what to unmap.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm/x86: Make sure we have the legacy x86 high memory reserved

Since there is a chance that the previous boot phase did not mark
this area as reserved through a memory region descriptor, make sure
we do it ourselved.

Since this memory region is very low, we do this insertion when this
list is empty to avoid unnecessary iterations.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm/x86: Make sure we allocate the `SIPI Vector` as reserved

Now that we can dynamically decide the `SIPI Vector`, make sure
we do this allocation when all the memory regions are properly
setup in the `Multiboot1` memory region setup method.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/common: Introduce `ukplat_memregion_list_insert_at_idx`

Implement the equivalent of `ukplat_memregion_list_insert` but one
that is aware of the index where it is inserting the memory region.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/common: Track index in `ukplat_memregion_list_insert_split_phys`

Make use of `ukplat_memregion_list_insert_at_idx` in order to make
insertion faster in `ukplat_memregion_list_insert_split_phys`.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/common/paging.c: Do not ignore `EEXIST` errors while mapping

Commit f0ae44de725f ("plat/kvm/x86: Workaround re-mapping low-mem areas")
introduced this check to ignore `EEXIST` errors while mapping the
memory regions marked as `UKPLAT_MEMRF_MAP` due to the fact that,
before the previous commit ("plat: Generalize memory initialization"),
Unikraft was ignoring all memory regions in the first 1 MiB and was
manually hardcoding the mappings in the static boot page tables.
This was done because the `multiboot` boot protocol would not report
the legacy mapped `BIOS ROM` area or the `VGA Framebuffer` in its
e820 map's reserved regions.

We no longer need this anymore and, if anything, it helps
us ensure that the setup of the boot environment's mappings is done
as we expect it to be.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/common/paging.c: For `ARM64`, don't map outside unmap region

Since on `ARM64` we map all of the peripherals' I/O space statically,
make sure that we do not map the regions in the list that are outside
our globally declared unmap memory region descriptor.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/linuxu: Add empty `bpt_unmap_mrd` memory region descriptors

In order not to generate undefined references when building and
not to unmap anything, add an empty `bpt_unmap_mrd` memory region
descriptor.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/xen/x86: Add `bpt_unmap_mrd` empty memory region descriptor

In order not to generate undefined references when building and
not to unmap anything, add an empty `bpt_unmap_mrd` memory region
descriptor.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

jschlumpp committed:
plat/kvm/x86: Insert legacy regions in linux boot entry

We use this region to write to the VGA console and Firecracker also
keeps this area clear. This ensures that the region is also mapped in
Unikraft.

Signed-off-by: Marco Schlumpp <marco@unikraft.io>

michpappas commented on plat/common/include/uk/plat/common/memory.h:
```diff
@@ -112,6 +112,39 @@ ukplat_memregion_list_insert(struct ukplat_memregion_list *list,
 }
 
+/**
+ * Inserts the region into the memory region list by index. The list is kept in
```
>  The list is kept in ascending order of physical base addresses. Order is only preserved among
> ranges that do not overlap. 

Inserting to index does not maintain the ordering by paddr, unless the caller knows what they're doing, so the description should be fixed. I see this is used for a very specific purpose in " plat/common: Track index in ukplat_memregion_list_insert_split_phys " where the caller knows what they're doing indeed.

Given that the requirement is that memregs must be in order and not overlapping, it would be best if it was made clear that this function should not be used unless strictly needed. For that it would be best if:
1. Update the description
2. Remove doxygen documentation

michpappas commented on plat/kvm/x86/setup.c:
```diff
@@ -220,7 +220,11 @@ static inline int cmdline_init(struct ukplat_bootinfo *bi)
 	}
 
-	cmdline = ukplat_memregion_alloc(cmdline_len + 1, UKPLAT_MEMRT_CMDLINE);
+	/* This is not the original command-line, but one that will be thrashed
```
What problem does this change fix? Is that only a change at the semantics level?

mogasergiu replied:
> What problem does this change fix? Is that only a change at the semantics level?

It bothered me how the type of this memory region was inconsistent with reality. This was not the command-line per se, but the command-line whose arguments will be separated by `\0`'s. This is inconsistent with the memory region that actually contains the command-line as a whole string. So I would like to keep this change. :)

michpappas replied:
Yes you're right. In fact I was also confused in the past with the two `cmdline` entries, so maybe this is a good change 👍🏼

michpappas commented on plat/common/memory.c:
```diff
@@ -137,3 +139,217 @@ void *ukplat_memregion_alloc(__sz size, int type)
+		mr_prio = get_mrd_prio(mr);
+
+		/* If they overlap */
+		if (RANGE_OVERLAP(ml->pbase,  ml->len, mr->pbase, mr->len)) {
```
This checks whether the regions overlap, but not whether they are part of the same mapping, which causes such regions not to be coalesced, which eventually manifests into a `-EXIST` error when the regions are mapped.

Fixing this here can actually be a bit tricky as you will need to know the granularity these regions are mapped with. If you decide to address this by allocating at least page-sized regions, you effectively change the requirements of the memregion API, as you will need to enforce that when allocating / adding new regions. CC: @marcrittinghaus.

Finally, notice that theoretically speaking this is still not fail-safe, as you have no guarantees that objects placed in memory by the bootloader (initrd, dtb, cmdline) are placed with granularity > `PAGE_SIZE`. I think in practice this is a non-issue, as boot protocols come with requirements for these objects (e.g. https://www.kernel.org/doc/Documentation/arm64/booting.txt), yet it's something worth to keep in mind.

mogasergiu replied:
I would solve this by allocating at least page-sized regions, as you have said. But, as you have also mentioned, you end up having to enforce this in the insertion API's as well, meaning that the previous boot phase that passes memory map related stuff will have to ensure that the raw received memory regions do not exist in the same mapping.

This is the solution that I would come up with now: a separate function, called something among the lines of `ukplat_memregion_list_page_overlap_adjust`, that should be called after `ukplat_memregion_list_coalesce`, that simply iterates through memory regions and removes `UKPLAT_MEMRF_MAP` if an in-page overlap is found. I would definetely not do this in the coalescing function as it would get out of its signature's scope.

So that is one suggestion. However, I still do not like it, as I believe that `ukplat_memregion_*` API's should have nothing to do with virtual memory or care about it as that kind of beats the purpose of having a paging API and a memregion API. I believe that `ukplat_memregion_*` API should only do what their names imply and take care of maintaining a proper raw memory map representation. Therefore, I think this case should be taken care of by `ukplat_paging_init`'s loop that maps memory regions with `UKPLAT_MEMRF_MAP`. Especially since it also aligns them to page-sized regions as well, making it already very easy to keep track of which page was mapped previously as they are already sorted ascendingly.

Let me know what you think of these alternate solutions. I would go with the latter.

michpappas replied:
> I would solve this by allocating at least page-sized regions

Just to be clear: multiple of page size , not at least page-sized

> Let me know what you think of these alternate solutions. I would go with the latter.

I understand the argumentation towards separation, but in practice I don't think you can ignore PT requirements until you reach the point you map, otherwise you could potentially end up with regions with conflicting attributes.

Consider for example the bug that triggered this discussion in the first place, where the `cmdl` and `dtb` regions were in the same page. The paging API could merge these into a single mapping if they had consistent protections and other PT attributes. But in any other case, you would end up with `-EEXIST`, which is what we are trying to avoid here. Because of that I see no choice other than enforcing regions to be a multiple of page size when they are created, and this is also the reason boot specs define alignment requirements on images. Similarly in the UEFI spec (e.g. Sect. 2.3.6 AArch64 Platforms - but similarly for other arch too):
> Any UEFI memory descriptor that requests a virtual mapping via the EFI_MEMORY_DESCRIPTOR having
the EFI_MEMORY_RUNTIME bit set must be aligned on a 4 KiB boundary and must be a multiple of 4 KiB
in size.

Given the above, I see no other way to address this problem. Unless I missed something in your proposal, that is :)

mogasergiu replied:
> otherwise you could potentially end up with regions with conflicting attributes.

Ah, very good point, yes! Right, I am only going to make sure this does not happen with `ukplat_memregion_alloc` (allocate page-aligned in multiple of page-sized) only though. I will leave the rest the same and only let this function keep this condition in mind (should also add a comment for this). This means I will also remove the code snippet that takes care of "in-image memory holes".

michpappas commented on plat/kvm/arm/entry64.S:
```diff
@@ -36,6 +36,13 @@
 #include <uk/config.h>
 #include <arch/reloc.h>
 
+/* Prefer using in-image stack, to avoid conflicts when unmapping is done by
```
How come we didn't come across this problem before? Not necessarily in this PR but it would be nice if we didn't waste a page for the bootstack.

mogasergiu replied:
I do not think a page is that much in our usual scenarios. The problem I encountered related to this line was 4 months ago, but from what I vaguely remember, it caused problems when setting up the memory regions, because I would have had to manually add a hardcoded separate function call just to insert the related memory region. Furthermore, at the time of doing this change I also did not have the `umap_bpt_mrd` idea, so a lot has changed since then. I will check again what happens if I do not do this, as I do not remember very well my intentions here.

mogasergiu replied:
@michpappas I just checked, the reason why I did this is because the paging initialization code, that also unmaps everything outside the kernel, uses the earliest boot stack and therefore ends up unmaps our stack and we end up using an unmapped stack while the paging initialization code is running. I will update the comment to be more verbose.

michpappas commented on plat/common/bootinfo.c:
```diff
@@ -6,6 +6,12 @@
+#include <libfdt.h>
+#endif /* CONFIG_LIBFDT */
+
+#define ukplat_bootinfo_crash(s)		ukplat_crash()
```
What is the purpose of this?

mogasergiu replied:
Just a wrapper like you would see in the initial Firecracker/Multiboot/UEFI code. Since console is not initialized at that point, I thought I would align this code with what has already been written for what would be executed before the console is initialized.

michpappas replied:
Okay, thanks for clarifying

michpappas commented on plat/common/bootinfo.c:
```diff
@@ -113,3 +119,142 @@ void ukplat_bootinfo_print(void)
+
+	fdt_bootinfo_fdt_mrd(bi, fdtp);
+	fdt_bootinfo_mem_mrd(bi, fdtp);
+	ukplat_memregion_list_coalesce(&bi->mrds);
```
As discussed: although not strictly required because `fdt_bootinfo_cmdl_mdr()` calls `ukplat_memregion_alloc()`, I think it's better form if you placed `ukplat_memregion_list_coalesce()` in the end, as it shows that you don't make any assumptions about the implementation of the various `fdt_bootinfo_*_mrd()` functions.

mogasergiu replied:
I just remembered why I did this. `fdt_bootinfo_cmdl_mrd` makes use of `ukplat_memregion_alloc` and, therefore, if we were to call `ukplat_memregion_list_coalesce` after it, it would imply that the list has already been coalesced/sorted before doing that allocation. So I will keep it like this and add a clarifying comment.

michpappas commented on plat/common/bootinfo.c:
```diff
@@ -113,3 +119,142 @@ void ukplat_bootinfo_print(void)
+		ukplat_bootinfo_crash("Could not find proper size cells!");
+
+	/*
+	 * QEMU will always provide us at least one bank of memory.
```
Is there a reason not to support >1 memory banks? IIRC the comment about QEMU supporting one bank originates from `plat/kvm/arm/setup.c`, yet it's time to get rid of the QEMU-specific mentality as Unikraft should be capable of running on any platform that provides a dtb on `arm64` (and this code is not part of `plat/qemu`). So if there is a blocker to support multiple banks, I would prefer if the comment was rephrased with the actual reasoning.

mogasergiu replied:
This code was just taken from the place you have already pointed to, I did not bother with improving it at that time. If I were to make this suggested change, it would have been in another commit. If you really want me to, I can update it to look for other memory banks as well, it should not be difficult and I would not mind it at all :).

michpappas replied:
Okay. Adding support for multiple banks would go together with excluding reserved memory. Perhaps to avoid delay merging this PR further we can add it later. Maybe just reduce the comment to something like "for now we only support one memory bank"?

michpappas commented on plat/common/bootinfo.c:
```diff
@@ -113,3 +119,142 @@ void ukplat_bootinfo_print(void)
+	const __u64 *regs;
+	int rc;
+
+	/* search for assigned VM memory in DTB */
```
This checks the validity of the header, so the comment and the error message are misleading (besides the fact that assigned and reserved memory are opposite things). Perhaps by this point checking against the header is redundant? See also the relevant comment in `fdt_bootinfo_fdt_mrd()`.

michpappas commented on plat/common/bootinfo.c:
```diff
@@ -113,3 +119,142 @@ void ukplat_bootinfo_print(void)
+
+	mrd.vbase = (__vaddr_t)mem_base;
+	mrd.pbase = (__paddr_t)mem_base;
+	mrd.len   = __BASE_ADDR - mem_base;
```
If the image is placed at the top of the region this will insert a region with zero length.

mogasergiu replied:
Very good point! I thought it would not matter since `ukplat_paging_init` ignores such regions and nothing would break. But yeah, the memory map ends up having an unnecessary memory region descriptor. Will add a check for it!

michpappas commented on plat/common/bootinfo.c:
```diff
@@ -113,3 +119,142 @@ void ukplat_bootinfo_print(void)
+
+	mrd.vbase = (__vaddr_t)__END;
+	mrd.pbase = (__paddr_t)__END;
+	mrd.len   = mem_base + mem_sz - __END;
```
If the image is placed at the end of the region this will insert a region of zero length.

michpappas commented on plat/common/bootinfo.c:
```diff
@@ -113,3 +119,142 @@ void ukplat_bootinfo_print(void)
+	if (unlikely(!regs || prop_len < prop_min_len))
+		ukplat_bootinfo_crash("Bad 'reg' property or more than one memory bank.");
+
+	mem_sz = fdt64_to_cpu(regs[1]);
```
Doesn't `fdt64_to_cpu()` imply 2 cells?

michpappas commented on plat/common/bootinfo.c:
```diff
@@ -113,3 +119,142 @@ void ukplat_bootinfo_print(void)
+	struct ukplat_memregion_desc mrd = {0};
+	int rc;
+
+	rc = fdt_check_header(fdtp);
```
Related to the comment in `fdt_bootinfo_mem_mrd()`: Perhaps you could move this check in `ukplat_bootinfo_fdt_setup()` after `ukplat_bootinfo_get()`.

michpappas commented on plat/common/paging.c:
```diff
@@ -741,7 +741,7 @@ static int pg_page_mapx(struct uk_pagetable *pt, __vaddr_t pt_vaddr,
 #endif /* CONFIG_PAGING_STATS */
 
-		if (PT_Lx_PTE_PRESENT(orig_pte, lvl))
+		if (PT_Lx_PTE_PRESENT(orig_pte, lvl) && pt == pg_active_pt)
```
Okay, yet it seems that this change is not relevant to this series which contains enough changes to review already. If that's the case, please move into a separate PR.

michpappas commented on plat/common/paging.c:
```diff
@@ -1418,3 +1421,101 @@ void ukplat_page_kunmap(struct uk_pagetable *pt, __vaddr_t vaddr,
+	return prot;
+}
+
+int ukplat_paging_init(void)
```
I think the right place to move this is `ukboot`,as the paging subsystem is (and should be) oblivious of `ukplat_memregion`. CC: @marcrittinghaus.

mogasergiu replied:
Yep, that would more sense for me as well, you are right :D.

EDIT: Nevermind, I think `ukplat_paging_init` should stay here until `setup.c`'s are merged. This is needed to make it easier to get rid of the BPT's while also being able to have early, pre-ukboot, access to important memory regions that need to be accessed early, like ACPI stuff. I believe this function should be part of the final, merged `setup.c`, when it will happen.

michpappas replied:
Yes exactly, this will be in the merging of `setup.c` logic into `ukboot`. On whether to keep it into `ukplat_paging_init()` until then, perhaps @marcrittinghaus can provide his feedback, being the author of the paging API.

michpappas commented on plat/common/memory.c:
```diff
@@ -380,3 +381,116 @@ int ukplat_memregion_get(int i, struct ukplat_memregion_desc **mrd)
+			&(struct ukplat_memregion_desc){
+				.vbase = ALIGN_UP(__END, __PAGE_SIZE),
+				.pbase = 0,
+				.len   = unmap_end -
```
Are there any conditions where any of the unmap regions has zero length?

mogasergiu replied:
`ukplat_page_unmap` ignores them and they are removed after `ukplat_paging_init` is done. So I think it is fine. I would keep it that way so I do not have to add a check when removing the two unmap regions.

michpappas commented on plat/common/paging.c:
```diff
@@ -1508,7 +1508,7 @@ int ukplat_paging_init(void)
 		rc = ukplat_page_map(&kernel_pt, vaddr, paddr,
 				     len >> PAGE_SHIFT, prot, 0);
-		if (unlikely(rc && rc != -EEXIST))
+		if (unlikely(rc))
```
Leaving this here as github does not allow commenting on the commit message:
```
before commit <generalize mem init commit sha here>, Unikraft was
```
You can't add a commit hash here even when the PR is finalized, because the hashes will changed by tags added by GitHub upon push.

mogasergiu replied:
I believe I wanted to have this commit in a separate PR, so that when it would have gotten merged I could have just amended the message. Oh well, guess I will just keep the commit message headline.

michpappas commented on plat/common/include/uk/plat/common/memory.h:
```diff
@@ -112,6 +112,32 @@ ukplat_memregion_list_insert(struct ukplat_memregion_list *list,
+#define X86_HI_MEM_LEN			0x50000
+
+static inline int
+ukplat_memregion_list_insert_legacy_hi_mem(struct ukplat_memregion_list *list)
```
Similarly for the following `SIPI` patch, these are too `x86_64` specific to be here. I understand that in the current state of the sources there may be not clear where these should be. According to the plat re-arch conventions that should probably be under `lib/ukboot/arch/x86_64/`.

michpappas commented on plat/common/Makefile.rules:
```diff
@@ -1,3 +1,15 @@
 	@$(SCRIPTS_DIR)/elf64_to_32.py $(1) -o $(1)
 endef
+
+BINFO_FLAGS := -a $(CONFIG_UK_ARCH)
```
According to plat re-arch these should go into the `Makefile.rules` of `lib/ukboot` so that they can be re-used by both internal and external platforms.

michpappas commented on plat/common/bootinfo.c:
```diff
@@ -10,8 +10,8 @@
  */
-__section(".uk_bootinfo") __align(8) __used
-static char bi_bootinfo_sec[UKPLAT_BOOTINFO_SIZE(32)];
+__section(".uk_bootinfo") __align(8) __used static char
```
I'm not very keen on the `fdt` code being in this file that contains generic functionality. By plat re-arch `bootinfo` will live under `lib/ukboot`, and different implementations that populate bootinfo like `multiboot`, `mbr`, `uefi` etc will be under `drivers/ukboot/xxx/`. Perhaps we should start moving things now instead of creating more work for re-arch?

mogasergiu replied:
Ok, I will have these placed in a `bootinfo_fdt.c` then.

michpappas commented on plat/common/memory.c:
```diff
@@ -91,6 +92,13 @@ void *ukplat_memregion_alloc(__sz size, int type, __u16 flags)
 		ostart = mrd->pbase;
 		olen   = mrd->len;
 
+		/* Check whether we are allocating from an in-image memory hole
```
Can you please provide some more context for this change? What is an example of such an allocation? What I read here is that if the allocation is outside the image, then we map. This makes the "hole" term confusing to me. Moreover, on the comment and in the description in the commit message you refer to "in-image holes" while the title is about "memory holes" in general.

mogasergiu replied:
Some ELF segments are separated by holes less than page-size, due to alignments and my coalesce function would mark those holes as free memory so that nothing goes to waste. If you allocate from there say a stack or cmdl through memregion_alloc, it would have MEMRF_MAP. And you get an EEXIST, because the ELF segments are mapped there already.

michpappas commented on plat/kvm/arm/bpt64.S:
```diff
@@ -203,6 +203,6 @@ arm64_bpt_l1_pt1:
 arm64_bpt_l0_pt0:
-	pte_fill	0x0000000040000000, 256, 0, PTE_PAGE_NORMAL_RO
-	pte_fill	0x0000000040100000, 256, 0, PTE_PAGE_NORMAL_RWX
+	pte_fill	0x0000000040000000, 512, 0, PTE_PAGE_NORMAL_RWX
```
Is this change in the scope of this commit? If so it should be described in the commit message.
Since we're there, also the commit message contains a bit too many "Furthermores", you may want to rephrase  :)

michpappas commented on plat/common/memory.c:
```diff
@@ -137,3 +139,217 @@ void *ukplat_memregion_alloc(__sz size, int type)
+				 * UKPLAT_MEMRT_KERNEL is of lower priority,
+				 * then something is definetely wrong => abort!
+				 */
+				if ((ml_prio < mr_prio &&
```
I don't quite follow this, even with the comment, can you please provide an example?

mogasergiu replied:
I am actually unsure whether this is needed anymore. I wrote this when writing UEFI for arm, at the beginning, before I was even met with the memregion missing for ARM... *sigh* :rofl:. At that time, ARM would set up its memory relative to ld symbols.

michpappas commented on plat/common/memory.c:
```diff
@@ -137,3 +139,217 @@ void *ukplat_memregion_alloc(__sz size, int type)
+				else
+					i++;
+
+			/* If they are of the same priority and different flags,
```
I'm a bit concerned about the silent keeping one region and deleting another. Would it be better to return an error unconditionally instead?

michpappas reviewed (changes requested):
@mogasergiu these changes are a big improvement! 🚀 Due to the sheer amount of changes I limited my review to common code and arm, ie I skipped linuxu, xen, and x86_64. Perhaps another reviewer can help with these.

You will notice that in several places I mention how the code should be laid out accoriding to the plat re-arch guidelines. Not sure if this should be done in the scope of this PR, perhaps we can discuss this offline.

michpappas reviewed (approved):
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
